arch/x86: add basic ACPI and non-trivial memory map support

gnuless opened the pull request:
These are building blocks for upcoming, more comprehensive x86 support.

The ACPI support is limited to the enumeration of local APICs from the MADT (read: available CPUs), though it can be extended as required in the future.  This is obviously of benefit in SMP scenario as-is, but also serves as a basis for other ACPI functions. (Unfortunately we can't hide from ACPI forever.)

This also implements a mechanism for dealing with the non-contiguous nature of RAM on even the simplest industry-standard x86 platform.  It supports retrieving memory maps from Multiboot if available, or a board/application-provided map, or defaults to a primitive map derived from devicetree.  It can later be extended to get information from EFI or other sources. This will undergird a more flexible page allocator to use in place of the current pool-based allocator in the existing x86 memory protection implementation.  Obviously it will have other uses (giving newlib a proper heap, for starters).

Included is a sample utility that demonstrates the use of these facilities.  The utility does a few other things useful for bringing up new boards (like empirical timer frequency measurement).

Commit:
arch/x86: rudimentary ACPI support

Implement a simple ACPI parser with enough functionality to
enumerate CPU cores and determine their local APIC IDs.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: multiboot: migrate multiboot initialization to early C

Originally, the multiboot info struct was copied in the early assembly
language code. This code is moved to a C function in multiboot.c for
two reasons:

1. It's about to get more complicated, as we want the ability to use
   a multiboot-provided memory map if available, and
2. this will faciliate its sharing between 32- and 64-bit subarches.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
samples/boards/x86_info: add platform information tool

This sample demonstrates basic use of the x86 multiboot and ACPI
systems, and also provides some useful information about the board
it's booted on: data handed over by the multiboot loader (which is
either QEMU or GRUB at this point), basic APIC CPU topology, and
timer driver frequency (computed empirically).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: add support for non-trivial memory maps

x86 has more complex memory maps than most Zephyr targets. A mechanism
is introduced here to manage such a map, and some methods are provided
to populate it (e.g., Multiboot).

The x86_info tool is extended to display memory map data.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

gnuless commented:
Rebased to deal with merge conflicts.

nashif commented on arch/x86/Kconfig:
```diff
@@ -80,6 +80,20 @@ choice
 		  Reboot via the RST_CNT register, going back to BIOS.
 endchoice
 
+config X86_ACPI
```
Can we just call it ACPI and make it depend on X86 or something? The X86 is either redundant (when you are adding this under arch/x86) and can't be reused if a feature appears on a different architecture, for example MULTIBOOT.

gnuless replied:
Hmm, well the Multiboot config options are all `X86_MULTIBOOT_*`, too. I'm just trying to be consistent. I figure if the day comes that we implement ACPI on ARM (and Itanium hahah) we'll probably want to reconcile the implementations with some common interface and *then* rename the CONFIG_X86_ACPI to just CONFIG_ACPI. Does that make sense?  Right now it only applies to x86 and is pretty X86-specific.

nashif replied:
I would call it CONFIG_ACPI now already, no reason to namespace it like this, if it is supported only by X86, just make a dependency on that. It is cleaner this way and future proof.

gnuless replied:
@nashif should I change X86_MULTIBOOT to MULTIBOOT as well, while I'm in here?

gnuless replied:
renamed

nashif commented on samples/boards/x86_info/prj.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_X86_ACPI=y
```
samples/boards/ is for actual boards, this is not a board specific sample, need a better place, maybe we should make a test out of this instead?

gnuless replied:
`tests/arch/x86/x86info`? I thought a sample was better because it's a utility rather more than a test, but I don't really care where it goes.

nashif replied:
yes, thats fine.

gnuless replied:
moved to `tests/arch/x86/info`

Commit:
arch/x86: rename CONFIG_X86_MULTIBOOT and related to CONFIG_MULTIBOOT

Simple naming change, since MULTIBOOT is clear enough by itself and
"namespacing" it to X86 is unnecessary and/or inappropriate.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: rename CONFIG_X86_ACPI and related to CONFIG_ACPI

ACPI is predominantly x86, and only currently implemented on x86,
but it is employed on other architectures, so rename accordingly.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
samples/board/x86_info: move to tests directory

This application's primary purpose is to provide some useful data to
the author of an x86 board support package for Zephyr-- it's not a
good sample.  It's not a good test either, but as a test it at least
prevents regressions in multiboot/ACPI builds.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

ceolin commented on include/arch/x86/multiboot.h:
```diff
@@ -51,15 +56,21 @@ extern void z_x86_multiboot_init(struct x86_multiboot_info *);
  */
 
+#define X86_MULTIBOOT_HEADER_FLAG_MEM	(1 << 1)  /* want mem_/mmap_* info */
+#define X86_MULTIBOOT_HEADER_FLAG_FB	(1 << 2)  /* want fb_* info */
```
nitpitck, use macro `BIT`. That macro uses proper unsigned type.

gnuless replied:
Fixed, thanks!

ceolin commented on samples/boards/x86_info/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2019 Intel Corp.
+
+cmake_minimum_required(VERSION 3.8.2)
```
all other places is setting 3.13.1. Don't know cmake enough to know if this version will be checked later or if it will explode when trying to use something from 3.13

gnuless replied:
Oh, wow. I'll fix this and file a bug - we have at handful of other tests/samples in the tree that set the wrong minimum (evidently I copied the boilerplate from one of those!).  

I seem to recall some weird thing with the way we use cmake that the minimum verison is ONLY enforced when it's in the top-level (in our case, app) `CMakeLists.txt`..

gnuless replied:
fixed!

Commit:
arch/x86: multiboot.h: use BIT() macros instead of explicit shifts

Because that's how we do it with Zephyr.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>
